### PR TITLE
add support for wrapped hyperdrive tokens

### PR DESCRIPTION
**Pull request by mcclurejt**
This enables support for hyperdrive instances with rebasing tokens since yearn is unable to support rebasing tokens.

### CODE REVIEW

**Comment by mcclurejt** [comment]
Leaving this in draft since I'm getting some very confusing approval reverts, will investigate more tomorrow

File: contracts/EverlongStrategyKeeper.sol
Comment (jalextowle) [inline contracts/EverlongStrategyKeeper.sol:342]: This comment is a bit confusing to me. Something like the following could be clearer since Hyperdrive doesn't have a token and actually has a base and vault shares token:

```
If the strategy's asset is a wrapped token, convert the spend amount to be denominated in units of vault shares.
```

Having said that, I'm not sure I understand what this code is doing. `canSpend` is initially set to:

```
uint256 canSpend = IERC20(IEverlongStrategy(_strategy).asset())
            .balanceOf(_strategy);
```

Isn't this already in units of the wrapped token since Everlong is holding something like wstETH in this case?

File: contracts/interfaces/IEverlongStrategy.sol
Comment (jalextowle) [inline contracts/interfaces/IEverlongStrategy.sol:90]: We talked about this before, but are these any different from `convertToBase` or `convertToShares`? I'll return to this once I finish my review, but, a priori, I'm not sure I see why they'd be any different.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:125]: "wrapped hyperdrive token" doesn't really make sense to me. wstETH isn't a wrapped Hyperdrive token, it's a wrapped variant of stETH.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:136]: Okay this is where the terminology is being introduced.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:674]: It seems like this is more of a nice to have refactor than a requirement. Unless I'm missing something, I'd consider reverting this change to reduce the number of lines that are changed by this PR.

I wouldn't be saying this if we were making this change before the audit, but now that we're in audits, it's best to keep the changes as small as possible.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:724]: Same here. This seems like a refactor rather than something that has to change. I'm not making this suggestion with `_openLong` and `_closeLong` since it seems like those functions are reducing the complexity.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:741]: Again, do we really need to be making this change now?

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:778]: Something to think about is that the minimum transaction amount is in units of base rather than vault shares. If the token is wrapped or if `asBase = false`, the `currentBalance` will be in units of vaults shares rather than base.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:801]: Nit: Let's add natspec to these functions.

Also, do we really need these? We can just use `convertToBase` for `convertToUnwrapped` and `convertToShares` for `convertToWrapped`. I'll leave this to you, but sticking with the old functions will reduce the lines of code.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:704]: Looks good.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:637]: The wrapped token is denominated in units of vault shares, not base tokens.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:609]: Nit: 

```suggestion
    /// @param _minVaultSharePrice Minimum hyperdrive vault share price to
    ///                purchase at.
```

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:602]: Okay, I see why we need this.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:533]:
```suggestion
                // Since this function would never be called as part of a
```

**Review by jalextowle — APPROVED** [review]
I'm going to go ahead and approve since I left all of the comments that I have. I know you said that you want to add some more testing, which I think would be a good addition in a follow-up. We should also talk to Sheng about fuzzing with rebasing tokens.

File: contracts/EverlongStrategyKeeper.sol
Comment (mcclurejt) [reply contracts/EverlongStrategyKeeper.sol:342]: You're totally correct! This can go :)

File: contracts/interfaces/IEverlongStrategy.sol
Comment (mcclurejt) [reply contracts/interfaces/IEverlongStrategy.sol:90]: They're not any different, the functions just add including a check for whether the strategy is using a wrapped token or not.

In terms of "understandability" I thought it'd be easier to follow using this language as opposed to `convertToBase` and `convertToShares`.

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:125]: Can see why this language is a bit confusing... intent here was that "hyperdrive token" means hyperdrive's base token or vault shares token consistent with the `hyperdriveToken` variable

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:674]: Totally on board with that

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:724]: totally on board

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:741]: +1

Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:741]: Actually I think this may include a critical change which is that calculateMaxLong doesn't convert to shares if the shares token is being used (i believe the returned amount is denominated in base).

Happy to include this in another PR though (or not at all and wait til after audit) whatever you think is best

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:778]: Great catch! Like the calcMaxLong issue should the fix for this be in a separate pr?

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:801]: Good catch on the natspec here....

While I understand it introduces a LOC increase, I feel as though it makes it far less confusing to use this language as opposed to `convertToShares` / `convertToBase`.

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:637]: Agree that the comment is confusing, but I think the logic is correct...

Correct, however the `_toSpend = _unwrap(_toSpend);` call above essentially calls `convertToBase` on the `_toSpend` which is needed for the approval, but must be reverted by calling `convertToWrapped` (aka `convertToShares`) . Doing this also comes with the benefit of conservative rounding to avoid off-by-ones

File: contracts/EverlongStrategy.sol
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:609]: I know it looks weird but this is actually consistent with how I've done commenting for parameters in the everlong repo (which I'm just now realizing is different than hyperdrive).

I like how hyperdrive does it but feel as though this should maybe be left as is and all parameter comments updated at once in a future pr.

**Comment by mcclurejt** [comment]
@jalextowle I created a separate PR for the conversion issues we were seeing (will be needed for tests to pass on this one)

https://github.com/delvtech/everlong/pull/38

Happy to do whatever you think is best with it

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [reply contracts/EverlongStrategy.sol:604]: Sounds good. Not a big deal, and I'm a fan of keeping things consistent.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [reply contracts/EverlongStrategy.sol]: Good call, I agree after thinking more about it.

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [reply contracts/EverlongStrategy.sol:738]: Sounds good

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [reply contracts/EverlongStrategy.sol]: We can include it here. I noticed the same thing elsewhere. Sorry for the miscommunication

File: contracts/interfaces/IEverlongStrategy.sol
Comment (jalextowle) [reply contracts/interfaces/IEverlongStrategy.sol:90]: Good call

File: contracts/EverlongStrategy.sol
Comment (jalextowle) [reply contracts/EverlongStrategy.sol]: Yep I agree. I was commenting on the comment rather than the logic. This is correct.

File: test/everlong/units/Wrapping.t.sol
Comment (brianmcmichael) [inline test/everlong/units/Wrapping.t.sol:97]: Since you're starting with a specific fork block number in VaultTest, you should be able to assert on an exact number consistently here.

**Pull request merged**
